File: backend/app/models/application.py
"""Application models - job applications and timeline events.

REQ-005 §4.5 - Application (Tier 4), TimelineEvent (Tier 5).
"""

import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    DateTime,
    ForeignKey,
    String,
    Text,
    UniqueConstraint,
    func,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, SoftDeleteMixin, TimestampMixin

_DEFAULT_UUID = text("gen_random_uuid()")
_ON_DELETE_SET_NULL = "SET NULL"

if TYPE_CHECKING:
    from app.models.cover_letter import CoverLetter, SubmittedCoverLetterPDF
    from app.models.job_posting import JobPosting
    from app.models.persona import Persona
    from app.models.persona_job import PersonaJob
    from app.models.resume import JobVariant, SubmittedResumePDF


class Application(Base, TimestampMixin, SoftDeleteMixin):
    """Job application record tracking status and materials.

    Tier 4 - references Persona, JobPosting, JobVariant, CoverLetter, SubmittedPDFs.
    """

    __tablename__ = "applications"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=_DEFAULT_UUID,
    )
    persona_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("personas.id", ondelete="CASCADE"),
        nullable=False,
    )
    job_posting_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("job_postings.id", ondelete="RESTRICT"),
        nullable=False,
    )
    job_variant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("job_variants.id", ondelete="RESTRICT"),
        nullable=False,
    )
    cover_letter_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("cover_letters.id", ondelete=_ON_DELETE_SET_NULL),
        nullable=True,
    )

    # PersonaJob link (REQ-015 §11 step 7)
    persona_job_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("persona_jobs.id", ondelete=_ON_DELETE_SET_NULL),
        nullable=True,
    )

    # Submitted PDF links (nullable for bidirectional FK - see REQ-005 §9.2)
    # use_alter=True handles circular dependency at migration level
    submitted_resume_pdf_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey(
            "submitted_resume_pdfs.id",
            ondelete=_ON_DELETE_SET_NULL,
            use_alter=True,
            name="fk_application_submitted_resume_pdf",
        ),
        nullable=True,
    )
    submitted_cover_letter_pdf_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey(
            "submitted_cover_letter_pdfs.id",
            ondelete=_ON_DELETE_SET_NULL,
            use_alter=True,
            name="fk_application_submitted_cover_letter_pdf",
        ),
        nullable=True,
    )

    # Frozen job snapshot at application time
    job_snapshot: Mapped[dict] = mapped_column(
        JSONB,
        nullable=False,
    )

    # Status tracking
    status: Mapped[str] = mapped_column(
        String(20),
        server_default=text("'Applied'"),
        nullable=False,
    )
    current_interview_stage: Mapped[str | None] = mapped_column(
        String(30),
        nullable=True,
    )

    # Outcome details
    offer_details: Mapped[dict | None] = mapped_column(
        JSONB,
        nullable=True,
    )
    rejection_details: Mapped[dict | None] = mapped_column(
        JSONB,
        nullable=True,
    )
    notes: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    # Pin control (REQ-012 Appendix A.1; archived_at inherited from SoftDeleteMixin)
    is_pinned: Mapped[bool] = mapped_column(
        Boolean,
        server_default=text("false"),
        nullable=False,
    )

    # Timestamps
    applied_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    status_updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    __table_args__ = (
        UniqueConstraint(
            "persona_id", "job_posting_id", name="uq_application_persona_job"
        ),
        CheckConstraint(
            "status IN ('Applied', 'Interviewing', 'Offer', 'Accepted', 'Rejected', 'Withdrawn')",
            name="ck_application_status",
        ),
        CheckConstraint(
            "current_interview_stage IN ('Phone Screen', 'Onsite', 'Final Round') OR current_interview_stage IS NULL",
            name="ck_application_interview_stage",
        ),
    )

    # Relationships
    persona: Mapped["Persona"] = relationship(
        "Persona",
        back_populates="applications",
    )
    job_posting: Mapped["JobPosting"] = relationship(
        "JobPosting",
        back_populates="applications",
    )
    job_variant: Mapped["JobVariant"] = relationship(
        "JobVariant",
        back_populates="applications",
    )
    persona_job: Mapped["PersonaJob | None"] = relationship(
        "PersonaJob",
        back_populates="applications",
    )
    # Note: CoverLetter.application is viewonly due to bidirectional FK pattern
    cover_letter: Mapped["CoverLetter | None"] = relationship(
        "CoverLetter",
        foreign_keys=[cover_letter_id],
    )
    submitted_resume_pdf: Mapped["SubmittedResumePDF | None"] = relationship(
        "SubmittedResumePDF",
        back_populates="application",
        foreign_keys="SubmittedResumePDF.application_id",
    )
    submitted_cover_letter_pdf: Mapped["SubmittedCoverLetterPDF | None"] = relationship(
        "SubmittedCoverLetterPDF",
        back_populates="application",
        foreign_keys="SubmittedCoverLetterPDF.application_id",
    )
    timeline_events: Mapped[list["TimelineEvent"]] = relationship(
        "TimelineEvent",
        back_populates="application",
        cascade="all, delete-orphan",
    )


class TimelineEvent(Base):
    """Event in an application's history timeline.

    Tier 5 - references Application.
    """

    __tablename__ = "timeline_events"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=_DEFAULT_UUID,
    )
    application_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("applications.id", ondelete="CASCADE"),
        nullable=False,
    )
    event_type: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
    )
    event_date: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
    )
    description: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )
    interview_stage: Mapped[str | None] = mapped_column(
        String(30),
        nullable=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    __table_args__ = (
        CheckConstraint(
            "event_type IN ('applied', 'status_changed', 'note_added', 'interview_scheduled', "
            "'interview_completed', 'offer_received', 'offer_accepted', 'rejected', 'withdrawn', "
            "'follow_up_sent', 'response_received', 'custom')",
            name="ck_timelineevent_event_type",
        ),
        CheckConstraint(
            "interview_stage IN ('Phone Screen', 'Onsite', 'Final Round') OR interview_stage IS NULL",
            name="ck_timelineevent_interview_stage",
        ),
    )

    # Relationships
    application: Mapped["Application"] = relationship(
        "Application",
        back_populates="timeline_events",
    )

File: frontend/src/types/application.ts
/**
 * Application & CoverLetter domain types matching
 * backend/app/models/application.py and backend/app/models/cover_letter.py.
 *
 * REQ-004: Application tracking pipeline.
 * REQ-005 §4.5: Database schema (Application, TimelineEvent).
 * REQ-005 §4.3: Database schema (CoverLetter, SubmittedCoverLetterPDF).
 * REQ-012 §11: Application tracking page.
 */

import type { WorkModel } from "./persona";

// ---------------------------------------------------------------------------
// Enum union types — match backend CHECK constraints
// ---------------------------------------------------------------------------

/** Backend: Application.status CHECK constraint. */
export type ApplicationStatus =
	| "Applied"
	| "Interviewing"
	| "Offer"
	| "Accepted"
	| "Rejected"
	| "Withdrawn";

/** Backend: Application.current_interview_stage CHECK constraint. */
export type InterviewStage = "Phone Screen" | "Onsite" | "Final Round";

/**
 * Backend: TimelineEvent.event_type CHECK constraint.
 *
 * Auto events: applied, status_changed, note_added, offer_received,
 *   offer_accepted, rejected, withdrawn.
 * Manual events: interview_scheduled, interview_completed,
 *   follow_up_sent, response_received, custom.
 */
export type TimelineEventType =
	| "applied"
	| "status_changed"
	| "note_added"
	| "interview_scheduled"
	| "interview_completed"
	| "offer_received"
	| "offer_accepted"
	| "rejected"
	| "withdrawn"
	| "follow_up_sent"
	| "response_received"
	| "custom";

/** Backend: CoverLetter.status CHECK constraint. */
export type CoverLetterStatus = "Draft" | "Approved" | "Archived";

/** Backend: ValidationIssue.severity from cover_letter_validation.py. */
export type ValidationSeverity = "error" | "warning";

// ---------------------------------------------------------------------------
// Enum value arrays — for form dropdowns and display
// ---------------------------------------------------------------------------

export const APPLICATION_STATUSES: readonly ApplicationStatus[] = [
	"Applied",
	"Interviewing",
	"Offer",
	"Accepted",
	"Rejected",
	"Withdrawn",
] as const;

export const INTERVIEW_STAGES: readonly InterviewStage[] = [
	"Phone Screen",
	"Onsite",
	"Final Round",
] as const;

export const TIMELINE_EVENT_TYPES: readonly TimelineEventType[] = [
	"applied",
	"status_changed",
	"note_added",
	"interview_scheduled",
	"interview_completed",
	"offer_received",
	"offer_accepted",
	"rejected",
	"withdrawn",
	"follow_up_sent",
	"response_received",
	"custom",
] as const;

export const COVER_LETTER_STATUSES: readonly CoverLetterStatus[] = [
	"Draft",
	"Approved",
	"Archived",
] as const;

export const VALIDATION_SEVERITIES: readonly ValidationSeverity[] = [
	"error",
	"warning",
] as const;

// ---------------------------------------------------------------------------
// JSONB sub-entity interfaces
// ---------------------------------------------------------------------------

/**
 * Frozen copy of job posting at application time.
 *
 * Backend: Application.job_snapshot JSONB column. Immutable after creation.
 * REQ-004 §4.1a: Job snapshot structure.
 */
export interface JobSnapshot {
	title: string;
	company_name: string;
	company_url: string | null;
	description: string;
	requirements: string | null;
	salary_min: number | null;
	salary_max: number | null;
	salary_currency: string | null;
	location: string | null;
	work_model: WorkModel | null;
	source_url: string | null;
	/** ISO 8601 datetime when snapshot was captured. */
	captured_at: string;
}

/**
 * Offer details captured when application status transitions to Offer.
 *
 * Backend: Application.offer_details JSONB column. All fields optional.
 * REQ-004 §4.3: Offer details structure.
 */
export interface OfferDetails {
	base_salary?: number;
	salary_currency?: string;
	bonus_percent?: number;
	equity_value?: number;
	equity_type?: "RSU" | "Options";
	equity_vesting_years?: number;
	/** ISO date string (YYYY-MM-DD). */
	start_date?: string;
	/** ISO date string (YYYY-MM-DD). Displayed as deadline countdown. */
	response_deadline?: string;
	other_benefits?: string;
	notes?: string;
}

/**
 * Rejection details captured when application status transitions to Rejected.
 *
 * Backend: Application.rejection_details JSONB column. All fields optional.
 * REQ-004 §4.4: Rejection details structure.
 */
export interface RejectionDetails {
	/** Stage at which rejection occurred. */
	stage?: string;
	reason?: string;
	feedback?: string;
	/** ISO 8601 datetime when rejection was communicated. */
	rejected_at?: string;
}

// ---------------------------------------------------------------------------
// Validation interfaces — match backend service dataclasses
// ---------------------------------------------------------------------------

/**
 * A single validation issue found in a cover letter draft.
 *
 * Backend: ValidationIssue dataclass (cover_letter_validation.py).
 * REQ-010 §5.4: Validation display.
 */
export interface ValidationIssue {
	/** "error" blocks presentation; "warning" is informational. */
	severity: ValidationSeverity;
	/** Machine-readable rule identifier (e.g., "length_min"). */
	rule: string;
	/** Human-readable description of the issue. */
	message: string;
}

/**
 * Result of validating a cover letter draft.
 *
 * Backend: CoverLetterValidation dataclass (cover_letter_validation.py).
 * REQ-010 §5.4: Validation display.
 */
export interface CoverLetterValidation {
	/** True if no errors (warnings allowed). */
	passed: boolean;
	/** Individual validation issues found. */
	issues: ValidationIssue[];
	/** Word count of the draft. */
	word_count: number;
}

// ---------------------------------------------------------------------------
// Main entity interfaces
// ---------------------------------------------------------------------------

/**
 * Job application record tracking status and materials.
 *
 * Backend: Application model (application.py). Tier 4 — references Persona,
 * JobPosting, JobVariant, CoverLetter, SubmittedPDFs.
 * REQ-004: Application tracking pipeline.
 * REQ-012 §11: Application tracking page.
 */
export interface Application {
	id: string;
	persona_id: string;
	job_posting_id: string;
	job_variant_id: string;
	cover_letter_id: string | null;
	submitted_resume_pdf_id: string | null;
	submitted_cover_letter_pdf_id: string | null;

	/** Frozen job posting copy at application time. */
	job_snapshot: JobSnapshot;

	// Status tracking
	status: ApplicationStatus;
	/** Sub-status when Interviewing. Null otherwise. */
	current_interview_stage: InterviewStage | null;

	// Outcome details
	offer_details: OfferDetails | null;
	rejection_details: RejectionDetails | null;
	/** Free-form notes, agent-populated or user-edited. */
	notes: string | null;

	// Pin control
	is_pinned: boolean;

	// Timestamps
	/** ISO 8601 datetime when user applied. */
	applied_at: string;
	/** ISO 8601 datetime of last status change. */
	status_updated_at: string;
	created_at: string;
	updated_at: string;
	/** ISO 8601 datetime. From SoftDeleteMixin. Null when not archived. */
	archived_at: string | null;
}

/**
 * Event in an application's history timeline.
 *
 * Backend: TimelineEvent model (application.py). Tier 5 — references Application.
 * Events are immutable (append-only, no edit/delete).
 * REQ-004 §5: Timeline events.
 */
export interface TimelineEvent {
	id: string;
	application_id: string;
	event_type: TimelineEventType;
	/** ISO 8601 datetime when the event occurred. */
	event_date: string;
	/** Event details. Null for auto-generated events with no extra info. */
	description: string | null;
	/** Interview stage associated with this event. Null for non-interview events. */
	interview_stage: InterviewStage | null;
	/** ISO 8601 datetime. Record creation time (no updated_at — events are immutable). */
	created_at: string;
}

/**
 * AI-generated cover letter for a job application.
 *
 * Backend: CoverLetter model (cover_letter.py). Tier 3 — references Persona,
 * JobPosting, Application (nullable).
 * REQ-002b: Cover letter generation.
 * REQ-010 §5/§7/§9: Cover letter validation and regeneration.
 */
export interface CoverLetter {
	id: string;
	persona_id: string;
	/** Null until application is created and linked. */
	application_id: string | null;
	job_posting_id: string;

	/** Achievement story UUIDs used in this letter. */
	achievement_stories_used: string[];

	// Content
	/** Editable while Draft status. */
	draft_text: string;
	/** Set on approval, immutable after. Null while Draft. */
	final_text: string | null;

	// Status
	status: CoverLetterStatus;

	/** Agent's story selection reasoning. Null if not yet generated. */
	agent_reasoning: string | null;

	/** Validation results from content checks. Null before validation runs. */
	validation_result: CoverLetterValidation | null;

	// Lifecycle timestamps
	/** ISO 8601 datetime. Set when status transitions to Approved. */
	approved_at: string | null;
	created_at: string;
	updated_at: string;
	/** ISO 8601 datetime. From SoftDeleteMixin. Null when not archived. */
	archived_at: string | null;
}

/**
 * Immutable PDF of cover letter submitted with a job application.
 *
 * Backend: SubmittedCoverLetterPDF model (cover_letter.py). Tier 4.
 * Binary content accessed via GET /submitted-cover-letter-pdfs/{id}/download.
 */
export interface SubmittedCoverLetterPDF {
	id: string;
	cover_letter_id: string;
	/** Null before application association. */
	application_id: string | null;
	file_name: string;
	/** ISO 8601 datetime. */
	generated_at: string;
}
